src/lib/backoff.ts

```diff
@@ -0,0 +1,88 @@
+import { promisify } from 'util';
+
+const sleep = promisify(setTimeout);
+
+export type BackoffCallback = {
+	retryCount: number;
+	delay: number;
+	error: any;
+};
+
+export type Options = {
+	attempt: number;
+	maxDelay: number;
+	minDelay: number;
+	maxRetries: number;
+	retryCallback?: (info: BackoffCallback) => void;
+};
+
+const DEFAULT_OPTIONS: Partial<Options> = {
+	attempt: 0,
+	maxDelay: 15000,
+	minDelay: 15000,
+	maxRetries: Infinity,
+};
+
+/**
+ * Retries a function with exponential delay between calls.
+ * If attempts is 0 then, it will delay for min(minDelay, maxDelay)
+ *
+ * Supports exceptions which return `retryAfter` value to specify the backoff duration.
+ * Will call `retryCallback` if passed in options as a way for callers to react to retries
+ */
+export function withBackoff<T extends (...args: any[]) => any>(
+	func: T,
+	options: Partial<Options> = {},
+) {
+	// If the function returns a promise, unwrap the promise return type
+	// otherwise use the actual return
+	type TReturn = ReturnType<T> extends Promise<infer R> ? R : ReturnType<T>;
+
+	const normalizedOptions: Options = {
+		...DEFAULT_OPTIONS,
+		...options,
+	} as Options;
+
+	return async function wrapped(...args: Parameters<T>): Promise<TReturn> {
+		let error: any;
```

Comment:
This is a nitpick but if you move the backoff logic to the catch clause you don't need this `let`

Reply:
I usually try to make try/catch blocks be as specific/small as possible but there is no other other to do after the try/catch so I'll move the backoff into the catch to remove initializing error.